File: woaa-frontend/src/api/trade.ts
/**
 * @fileoverview Handles API requests related to trades.
 */

const BASE_URL = import.meta.env.VITE_API_URL;

//TODO: add log for all of them

/**
 * Create a new trade for the authenticated user.
 *
 * Payload structure:
 * {
 *   symbol: string,           // Stock ticker, e.g. "AAPL"
 *   shares: number,           // Number of shares (> 0)
 *   price: number,            // Trade price per share (> 0)
 *   action: "buy" | "short",  // Trade direction
 *   notes?: string            // Optional notes
 * }
 */
export async function createTrade(trade: any): Promise<any> {
  const token = localStorage.getItem("token");
  const res = await fetch(`${BASE_URL}/trades/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(trade),
  });

  // 🔹 If the response is not OK, parse the error and throw it
  if (!res.ok) {
    let errorMsg = res.statusText;
    try {
      const errData = await res.clone().json();
      errorMsg = errData.detail || errData.message || res.statusText;
    } catch {
      /* ignore JSON parse errors, fallback to statusText */
    }
    throw new Error(errorMsg);
  }

  return res.json();
}

/**
 * Close an existing position (sell for Long, cover for Short).
 *
 * @param positionId - ID of the position to close
 * @param currentPrice - execution price (> 0)
 * @param notes - optional notes
 */
export async function closeTrade(
  positionId: string,
  currentPrice: number,
  notes: string = ""
): Promise<any> {
  const token = localStorage.getItem("token");

  const res = await fetch(
    `${import.meta.env.VITE_API_URL}/trades/item/${positionId}/close`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ current_price: currentPrice, notes }),
    }
  );

  // TODO: add log

  if (!res.ok) {
    let errorMsg = res.statusText;
    try {
      const errData = await res.clone().json();
      errorMsg = errData.detail || errData.message || res.statusText;
    } catch {
      /* ignore JSON parse errors */
    }
    throw new Error(errorMsg);
  }

  return res.json();
}

/**
 * Delete (power-up delete) an open position.
 * Charges fee (per user settings), refunds principal to sim, deletes position.
 *
 * @param positionId - ID of the position to delete
 */
export async function deleteTrade(positionId: string): Promise<any> {
  const token = localStorage.getItem("token");

  const res = await fetch(`${BASE_URL}/trades/item/${positionId}`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  // TODO: add log

  if (!res.ok) {
    let errorMsg = res.statusText;
    try {
      const errData = await res.clone().json();
      errorMsg = errData.detail || errData.message || res.statusText;
    } catch {
      /* ignore JSON parse errors */
    }
    throw new Error(errorMsg);
  }

  // Handle both 200 with JSON body and 204 No Content
  if (res.status === 204) return { ok: true };
  try {
    return await res.json();
  } catch {
    return { ok: true };
  }
}

File: woaa-backend/app/api/trade.py
# app/api/trade.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from uuid import UUID

from app.database import get_db
from app.auth import get_current_user
from app.models.user import User

from app.services.trade import close_trade_service, create_trade_service, delete_trade_service  # ✅ import service

router = APIRouter(prefix="/trades", tags=["trades"])


@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_trade(
    payload: dict,  # Expected keys:
                    # {
                    #   "symbol": str,          # Stock ticker, e.g. "AAPL"
                    #   "shares": float,        # Number of shares (> 0)
                    #   "price": float,         # Trade price per share (> 0)
                    #   "action": "buy" | "short",  # Trade direction
                    #   "notes": str (optional) # Optional trade notes
                    # }
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Create (open) a trade for the authenticated user.
    """
    return await create_trade_service(
        db=db,
        user_id=current_user.id,
        trade_data=payload
    )


@router.post("/item/{position_id}/close", status_code=status.HTTP_200_OK)
async def close_trade(
    position_id: UUID,
    payload: dict,  # Expected keys:
                    # {
                    #   "current_price": float,   # required, > 0
                    #   "notes": str (optional)   # optional notes
                    # }
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Close an existing open position belonging to the authenticated user."""
    # Minimal inline validation since we're not using a schema
    if "current_price" not in payload:
        raise HTTPException(status_code=400, detail="Missing field: current_price")
    current_price = payload["current_price"]
    if not isinstance(current_price, (int, float)) or current_price <= 0:
        raise HTTPException(status_code=400, detail="current_price must be a positive number")

    notes = payload.get("notes", "")

    return await close_trade_service(
        db=db,
        user_id=current_user.id,
        position_id=position_id,
        current_price=float(current_price),
        notes=notes,
    )


@router.delete("/item/{position_id}", status_code=status.HTTP_200_OK)
async def delete_trade(
    position_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Power-up delete:
      - charges fee (real vs sim per user settings)
      - refunds principal to sim
      - deletes the position only
      - returns updated balances and summary
    """
    return await delete_trade_service(
        db=db,
        user_id=current_user.id,
        position_id=position_id,
    )